### PR TITLE
Fix action space check

### PR DESCRIPTION
**Pull request by pbrubeck**
The space check in `action` was not quite right, it is hard to tell exactly why. Previously it only passed because `V == V.dual()` in Firedrake.

This PR removes the need to check specific types of the arguments and handles them in a more symmetric way.

### CODE REVIEW

**Comment by pbrubeck** [comment]
This ruff thing is ridiculous

File: ufl/action.py [hunk 1]
```diff
@@ -159,29 +157,36 @@ def __hash__(self):
 
 def _check_function_spaces(left, right):
```
Comment (pbrubeck) [inline ufl/action.py:158]: Dear code reviewers,

Please ignore the rest of the diff, that's just ruff.
Only take a look at this function.

Thank you.

**Comment by pbrubeck** [comment]
@jorgensd FeniCSx CI fails, but I suspect it is for an unrelated reason?

**Comment by jorgensd** [comment]
> @jorgensd FeniCSx CI fails, but I suspect it is for an unrelated reason?

@finsberg, it seems like you missed a test with your latest PR:
https://github.com/FEniCS/dolfinx/blob/0c3a33ec5cfc5fef3e5bc965fcd3df533ed4aee6/python/test/unit/io/test_xdmf_function.py#L232-L249

**Review by dham — APPROVED** [review]
I think this looks right.

**Pull request merged**